src/consumption.cpp

```diff
@@ -205,191 +217,159 @@ morale_type player::allergy_type( const item &food ) const
-        }
+        consequences.emplace_back( ALLERGY,
+                                   _( "You still feel nauseous and will probably puke it all up again." ) );
     }
 
-    // Print at most one of those
-    bool overfull = false;
-    if( overeating ) {
-        overfull = !maybe_query( _( "You're full.  Force yourself to eat?" ) );
-    } else if( ( ( nutr > 0 && temp_hunger < capacity ) ||
-                 ( comest->quench > 0 && temp_thirst < capacity ) ) &&
-               !food.has_infinite_charges() ) {
-        // This first section is only for flavor msg, skip if we wouldn't print anything
-        if( interactive &&
-            ( slimespawner ||
-              ( gourmand && has_active_mutation( trait_id( "GOURMAND" ) ) ) ||
-              ( eathealth && has_active_mutation( trait_id( "EATHEALTH" ) ) ) ) ) {
-            add_msg_if_player( _( "You're full, but you cram it into your mouth without a second thought." ) );
-        } else {
-            overfull = !maybe_query( _( "You will not be able to finish it all.  Consume it?" ) );
+    const int nutr = nutrition_for( food.type );
+    const int quench = comest->quench;
+    const int temp_hunger = get_hunger() - nutr;
+    const int temp_thirst = get_thirst() - quench;
+
+    if( has_active_mutation( trait_id( "HIBERNATE" ) ) ) {
+        if( ( get_hunger() >= -60 && get_thirst() >= -60 ) &&
+            ( temp_hunger < -60 || temp_thirst < -60 ) ) {
+            consequences.emplace_back( TOO_FULL, _( "You're adequately fueled. Prepare for hibernation?" ) );
+        }
+    } else if( !has_active_mutation( trait_id( "EATHEALTH" ) ) &&
+               !has_trait( trait_id( "SLIMESPAWNER" ) ) ) {
+        if( get_hunger() < 0 && nutr >= 5 && !has_active_mutation( trait_id( "GOURMAND" ) ) ) {
+            consequences.emplace_back( TOO_FULL, _( "You're full already." ) );
+        } else if( ( ( nutr > 0 && temp_hunger < stomach_capacity() ) ||
+                     ( comest->quench > 0 && temp_thirst < stomach_capacity() ) ) &&
+                   !food.has_infinite_charges() ) {
+            consequences.emplace_back( TOO_FULL, _( "You will not be able to finish it all." ) );
         }
     }
 
-    if( overfull ) {
-        return TOO_FULL;
+    const int energy = get_acquirable_energy( food );
+
+    if( energy ) {
+        consequences.emplace_back( INEDIBLE,
+                                   string_format( _( "Instead of eating it, you could get %d points of energy." ),
+                                           energy ) );
     }
 
+    if( !consequences.empty() ) {
+        const auto res = consequences.front().first;
+        if( !interactive ) {
+            return res;
+        }
+        std::ostringstream req;
+        for( const auto &elem : consequences ) {
+            req << elem.second << std::endl;
+        }
+        req << _( "Consume anyway?" );
+
+        if( !query_yn( "%s", req.str().c_str() ) ) {
+            return res;
```

Comment:
I have reordered the consequences by severity (the order is arguable):

Type | Consequences
--- | ---
Rotten		| Food poisoning, health damage
Cannibalism	| Long lasting morale penalty
Nausea		| Loosing eaten food, hunger, thirst, morale penalty
Allergy		| Morale penalty that lasts half as long as for cannibalism
Too full		| Risk of sudden vomiting
Won't eat	| No ill effects